File: aiworkbench_api-bugfix_735_api (1)/aiworkbench_api-bugfix_735_api/app/core/security_headers.py
"""Shared HTTP security header values for API responses."""

from __future__ import annotations

from app.core.config import settings

HSTS_MAX_AGE_SECONDS = 31_536_000  # 1 year

API_CONTENT_SECURITY_POLICY = (
    "default-src 'none'; "
    "frame-ancestors 'none'; "
    "base-uri 'none'; "
    "form-action 'none'; "
    "object-src 'none'"
)


def build_api_security_headers(*, include_hsts: bool | None = None) -> dict[str, str]:
    """Return security headers suitable for JSON API responses."""
    if include_hsts is None:
        include_hsts = settings.ENVIRONMENT == "production"

    headers = {
        "X-Content-Type-Options": "nosniff",
        "X-Frame-Options": "DENY",
        "Referrer-Policy": "strict-origin-when-cross-origin",
        "Permissions-Policy": "camera=(), microphone=(), geolocation=(), payment=()",
        "Content-Security-Policy": API_CONTENT_SECURITY_POLICY,
        "Cache-Control": "no-store, no-cache, must-revalidate, private",
        "Pragma": "no-cache",
    }

    if include_hsts:
        headers["Strict-Transport-Security"] = f"max-age={HSTS_MAX_AGE_SECONDS}; includeSubDomains"

    return headers

File: aiworkbench_ui-bugfix_735_ui/aiworkbench_ui-bugfix_735_ui/src/config/securityHeaders.ts
/** HTTP security headers for the SPA (dev server, preview, and nginx reference). */

const HSTS = 'max-age=31536000; includeSubDomains';

export const PRODUCTION_CSP = [
  "default-src 'self'",
  "script-src 'self'",
  "style-src 'self' 'unsafe-inline' https://fonts.googleapis.com",
  "img-src 'self' data: blob: https:",
  "font-src 'self' data: https://fonts.gstatic.com",
  "connect-src 'self'",
  "worker-src 'self' blob:",
  "frame-ancestors 'self'",
  "object-src 'none'",
  "base-uri 'self'",
  "form-action 'self'",
  "upgrade-insecure-requests",
].join('; ');

export const DEVELOPMENT_CSP = [
  "default-src 'self'",
  "script-src 'self' 'unsafe-inline' 'unsafe-eval'",
  "style-src 'self' 'unsafe-inline' https://fonts.googleapis.com",
  "img-src 'self' data: blob: https:",
  "font-src 'self' data: https://fonts.gstatic.com",
  "connect-src 'self' ws: wss:",
  "worker-src 'self' blob:",
  "frame-ancestors 'self'",
  "object-src 'none'",
  "base-uri 'self'",
  "form-action 'self'",
].join('; ');

export function buildSpaSecurityHeaders(options: { development?: boolean; includeHsts?: boolean } = {}) {
  const development = options.development ?? false;
  const includeHsts = options.includeHsts ?? !development;

  const headers: Record<string, string> = {
    'X-Content-Type-Options': 'nosniff',
    'X-Frame-Options': 'SAMEORIGIN',
    'Referrer-Policy': 'strict-origin-when-cross-origin',
    'Permissions-Policy': 'camera=(), microphone=(), geolocation=(), payment=()',
    'Content-Security-Policy': development ? DEVELOPMENT_CSP : PRODUCTION_CSP,
  };

  if (includeHsts) {
    headers['Strict-Transport-Security'] = HSTS;
  }

  return headers;
}

export function buildHtmlCacheControl(pathname: string): string {
  if (pathname === '/' || pathname.endsWith('.html')) {
    return 'no-store, no-cache, must-revalidate, private';
  }
  if (/\.(?:js|css|png|jpg|jpeg|webp|gif|svg|woff2?|ico)$/.test(pathname)) {
    return 'public, max-age=31536000, immutable';
  }
  return 'no-store, no-cache, must-revalidate, private';
}
